### PR TITLE
Quantization aware training: Freeze batch norm support

### PR DESCRIPTION
**Pull request by raghuramank100**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26627 Default observer and fake-quant for backends
* #26625 Emulate weight and activation only quant with fake quant, numerics test
* **#26624 Quantization aware training: Freeze batch norm support**
* #26623 Per channel fake quant

For QAT we need to be able to control batch norm for all modules from the top. Adding helper functions to enable/disable batch norm freezing during training

Differential Revision: [D17512199](https://our.internmc.facebook.com/intern/diff/D17512199/)

### CODE REVIEW

File: torch/nn/_intrinsic/qat/modules/conv_fused.py [hunk 1]
```diff
@@ -219,6 +227,14 @@ def __init__(self,
     def forward(self, input):
         return self.observer(F.relu(super(ConvBnReLU2d, self)._forward(input)))
 
+    def update_bn_stats(self):
```
Comment (jerryzh168) [inline torch/nn/_intrinsic/qat/modules/conv_fused.py]: this module inherits from ConvBn2d, do we still need this function?

File: torch/nn/_intrinsic/qat/modules/conv_fused.py [hunk 2]
```diff
@@ -261,3 +277,11 @@ def forward(self, input):
     @classmethod
     def from_float(cls, mod, qconfig=None):
         return super(ConvReLU2d, cls).from_float(mod, qconfig)
+
+def update_bn_stats(mod):
+    if type(mod) in set([ConvBnReLU2d,ConvBn2d]):
```
Comment (jerryzh168) [inline torch/nn/_intrinsic/qat/modules/conv_fused.py]: formatting

**Review by dzhulgakov — COMMENTED** [review]
General question, probably not for this diff - do we need freeze_bn support also in non-fused modules?

**Comment by raghuramank100** [comment]
> General question, probably not for this diff - do we need freeze_bn support also in non-fused modules?

Setting bn to eval() would do that and allow for frozen statistics to be used during training, but this is ugly as we want the rest of the modules to be in train(), with bn alone being in eval.
I think we need this supported in bn so that we have an independent way to control this outside of .train() and .eval(). An example for this need is the frozenbatchnorm module in torchvision: https://github.com/pytorch/vision/blob/master/torchvision/ops/misc.py#L135

File: torch/nn/_intrinsic/qat/modules/conv_fused.py [hunk 3]
```diff
@@ -251,3 +259,11 @@ def forward(self, input):
     @classmethod
     def from_float(cls, mod, qconfig=None):
         return super(ConvReLU2d, cls).from_float(mod, qconfig)
+
+def update_bn_stats(mod):
+    if type(mod) in set([ConvBnReLU2d, ConvBn2d]):
```
Comment (dzhulgakov) [inline torch/nn/_intrinsic/qat/modules/conv_fused.py]: you can also just do 'hasattr' (in case more modules appear in the future)

**Review by dzhulgakov — APPROVED** [review]

File: torch/nn/_intrinsic/qat/modules/conv_fused.py (hunk 1 above)
Comment (raghuramank100) [reply torch/nn/_intrinsic/qat/modules/conv_fused.py]: Good point, have moved this to the base class (convBn2d) and removed it in convbnrelu2d

File: torch/nn/_intrinsic/qat/modules/conv_fused.py (hunk 3 above)
Comment (raghuramank100) [reply torch/nn/_intrinsic/qat/modules/conv_fused.py]: Is this safer? We want to modify this field only for specific modules. Tagging based on a specific attribute could be theoretically more risky as we would set freeze_bn to false for any module that has this field. Leaving this as is so that we have explicit control over which modules we apply this modification to.
